api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -238,6 +238,52 @@ message HttpConnectionManager {
     google.protobuf.BoolValue enabled = 3;
   }
 
+  // Transformations that apply to path headers. The policy to apply these
+  // transformations can be tuned by the configuration field. Transformations
+  // can be applied internally only, so that the normalized header is used in
+  // matching, routing, and HTTP filters while the original path header is
+  // forwarded upstream, or additionally apply to the forwarded *:path* header.
+  message PathNormalizationOptions {
+    // Defines a policy to apply normalization.  Path rewrites :ref:`regex_rewrite
+    // <envoy_api_field_route.RouteAction.regex_rewrite>` or :ref:`prefix_rewrite
+    // <envoy_api_field_route.RouteAction.prefix_rewrite>` will apply to the
+    // *:path* header destined for the upstream.
+    enum NormalizationPolicy {
+      // Normalizations apply internally before any processing of requests by HTTP
+      // filters, routing, and matching, *and* will affect the upstream *:path* header
+      // as well.
+      FORWARD = 0;
+
+      // Normalizations only apply internally. The original *:path* header will
+      // be sent upstream *unless* the path has been mutated by a filter. Any
+      // regex rewrites that the router performs will be on the upstream *:path*
+      // header.
+      INTERNAL = 1;
```

Comment:
I wonder if there is a case for being able to orthogonally specify both the match and forward behavior? E.g. someone wants to be able to do RFC3986 on both but only merge_slashes on match? It seems a bit contrived, but as we add more normalizations, maybe it becomes interesting. I think code/API wise it's not hard to support this dual config.

Reply:
Yes, I did consider this, especially with the mindset of making this as general as possible. It comes with the downside of having to specify the behavior for each transformation. I wonder if users can either specify a catch all policy OR a policy for each transformation if they want a very customizeable experience. Something like:

PathNormalizationOptions:
    - merge_slashes: true
      policy: internal
    - path_normalization: true
      policy: forward

PathNormalizationOptions:
    - uniform_policy: forward
    - merge_slashes: true
    - path_normalization: true